cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -1415,22 +1435,69 @@ pub(crate) fn lower_icmp<C: LowerCtx<I = Inst>>(
                 }
             }
         }
+        condcode
     } else if !ty.is_vector() {
-        let alu_op = choose_32_64(ty, ALUOp::SubS32, ALUOp::SubS64);
         let rn = put_input_in_reg(ctx, inputs[0], narrow_mode);
         let rm = put_input_in_rse_imm12(ctx, inputs[1], narrow_mode);
+
+        let is_overflow = condcode == IntCC::Overflow || condcode == IntCC::NotOverflow;
+        let is_small_type = ty == I8 || ty == I16;
+        let (out_condcode, rn, rm) = if is_overflow && is_small_type {
+            // Overflow checks for non native types require additional instructions, other than
+            // just the extend op.
+            //
+            // TODO: Codegen improvements: Merge the second sxt{h,b} into the following sub instruction.
+            //
+            // sxt{h,b}  w0, w0
+            // sxt{h,b}  w1, w1
+            // sub       w0, w0, w1
+            // cmp       w0, w0, sxt{h,b}
+            //
+            // This comparision leaves the result in the `ne` flag, so we need to signal that
+            // to the caller
+
+            let extend_op = if ty == I8 {
+                ExtendOp::SXTB
+            } else {
+                ExtendOp::SXTH
+            };
+            let tmp1 = ctx.alloc_tmp(I32).only_reg().unwrap();
+            ctx.emit(alu_inst_imm12(ALUOp::Sub32, tmp1, rn, rm));
+
+            let out_condcode = match condcode {
+                IntCC::Overflow => IntCC::NotEqual,
+                IntCC::NotOverflow => IntCC::Equal,
+                _ => unreachable!(),
+            };
+            (
+                out_condcode,
+                tmp1.to_reg(),
+                ResultRSEImm12::RegExtend(tmp1.to_reg(), extend_op),
+            )
+        } else {
+            (condcode, rn, rm)
+        };
+
+        let alu_op = choose_32_64(ty, ALUOp::SubS32, ALUOp::SubS64);
         ctx.emit(alu_inst_imm12(alu_op, writable_zero_reg(), rn, rm));
 
         if let IcmpOutput::Register(rd) = output {
```

Comment:
I think that there should be just a single block like this one, right before the function returns.

Reply:
We don't do this everywhere, namely I128 gt/lt/etc and vector icmps, so I don't know how well this ends up being in practice. But I pushed a commit with this at the end, let me know which way you prefer it.

Reply:
Are you saying that there are cases in which we pass `IcmpOutput::Flags`, but we return the result in a register (or vice versa)?

Reply:
Thanks, that is much better!

Reply:
Oh, right, a bit of tunnel vision on my side - I was absolutely concentrating on the flags case.